[larainezo] iP

larainezo opened the pull request:
### Duchess

> 🤨Can you really keep track of everything by memory?🤨

Stop overestimating your memory potential and let Duchess be your personal task manager!

**Here is what she can do for you:**
- Keep track of your tasks
- Auto-check for incorrect input (_people do make mistakes and that's fine!_)

Types of tasks:
1. To Do
2. Deadline
3. Event

To use it:
- [ ] Download the file [here](https://github.com/larainezo/ip)
- [ ] Run the following command in your terminal
`java -jar Duchess.jar`

What are you waiting for? It is now on offer at ~~$100~~ **$0**

🎉@larainezo 

Also, in case you need to modify the program to your preference, I have made it simple through OOP!
```java
public Duchess() {
        ui = new Ui();
        storage = new Storage("tasks.txt");
        parser = new Parser();
        try {
            tasks = new TaskList(storage.load());
        } catch (DuchessException e) {
            ui.showLoadingError();
            tasks = new TaskList();
        }
    }
``` 

Version currently in Space Grey `#717378`

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

larainezo committed:
Level-0

larainezo committed:
Level-1

larainezo committed:
Level-2

larainezo committed:
Level-3

larainezo committed:
Level-4

larainezo committed:
A-TextUiTesting

larainezo committed:
A-Exceptions

larainezo committed:
Level-6

larainezo committed:
Level-5

larainezo committed:
Level-6

larainezo committed:
Level-7

larainezo committed:
Revert "Level-7"

This reverts commit 18c1fd06ed6919fa7c7c5e415d2c22667d2bc287.

larainezo committed:
Level-7

larainezo committed:
Revert "Level-7"

This reverts commit b51ff49ec5f2377097bc80f7363b1836ec671e0c.

larainezo committed:
Level-7

larainezo committed:
Level-7

larainezo committed:
Level-7

larainezo committed:
Add save feature

larainezo committed:
Merge pull request #1 from larainezo/branch-Level-7

Merge branch-Level-7

larainezo committed:
Change task dates from String to LocalDate format

larainezo committed:
Merge pull request #2 from larainezo/branch-Level-8

Change task dates from String to LocalDate format

larainezo committed:
Add more OOP classes - Ui, Storage, Parser, TaskList

larainezo committed:
Merge pull request #3 from larainezo/A-MoreOOP

A-MoreOOP

larainezo committed:
Create duchess, duchess.util, duchess.tasks packages

larainezo committed:
Merge pull request #4 from larainezo/A-Packages

A-Packages

larainezo committed:
Merge pull request #5 from larainezo/add-gradle-support

A-Gradle

larainezo committed:
Add JUnit tests for duchess.tasks.Task and duchess.util.TaskList

larainezo committed:
Merge pull request #6 from larainezo/A-JUnit

A-JUnit

larainezo committed:
Create Duchess.jar executable

larainezo committed:
Merge pull request #7 from larainezo/A-Jar

A-Jar

larainezo committed:
Add JavaDoc for all classes

larainezo committed:
Add JavaDoc for all classes

larainezo committed:
Fix if-else in Parser.java according to Java coding standard

larainezo committed:
Add find feature to find tasks based on keyword

larainezo committed:
Add find feature to find tasks based on keyword

larainezo committed:
Merge pull request #8 from larainezo/A-JavaDoc

A-JavaDoc

larainezo committed:
Merge pull request #9 from larainezo/A-CodingStandard

A-CodingStandard

larainezo committed:
Merge branch 'master' into Level-9

larainezo committed:
Merge pull request #10 from larainezo/Level-9

Level-9

larainezo committed:
Add missing getDescription() function to Task class

larainezo committed:
Add Checkstyle with Gradle

larainezo committed:
Merge pull request #11 from larainezo/A-CheckStyle

A-CheckStyle

javierng2knus commented on src/main/java/duchess/tasks/Task.java:
```diff
@@ -0,0 +1,59 @@
+
+public class Task {
+    protected String description;
+    protected boolean isDone;
```
Great job for following the boolean naming convention 👍

javierng2knus commented on src/main/java/duchess/util/TaskList.java:
```diff
@@ -0,0 +1,127 @@
+        return this.taskList;
+    }
+
+    /**
+     * Delete Task object in ArrayList of Task objects.
+     *
+     * @param id id of task in list to be deleted.
+     */
+    public void deleteTask(int id) throws IndexOutOfBoundsException{
+        String lineBreak = "\t\t------------------------------------------";
+        Task t = taskList.remove(id - 1);
+        System.out.println(lineBreak);
+        System.out.println("\t\tNoted. I've removed this task:");
+        System.out.println("\t\t  " + t);
+        System.out.println("\t\tNow you have " + taskList.size() + " tasks in the list.");
+        System.out.println(lineBreak);
+    }
```
I like how you included Javadoc comprehensively! 👍 
There's only one missing `@throw` for this function, but other than that, solid work 👌

javierng2knus commented on src/main/java/duchess/util/Storage.java:
```diff
@@ -0,0 +1,112 @@
+
+                    String[] input = data.split("");
+                    String type = input[1];
+                    String isDone = input[4];
```
Would it be better if a different name was used, instead of naming the string as `isDone`
Perhaps using `markedStatus` or `doneChar` would be clearer?

Or maybe keeping the name but instead convert it to a boolean using `input[4].equals('X') ? true : false`

Please feel free to share your thoughts regarding my suggestions! 😊

javierng2knus commented on src/main/java/duchess/util/TaskList.java:
```diff
@@ -0,0 +1,127 @@
+     *
+     * @param id id of task in list to be deleted.
+     */
+    public void deleteTask(int id) throws IndexOutOfBoundsException{
```
It seems for this function, you may have missed a white space after the `IndexOutOfBoundsException`

There appears to be one other location, in the `addToDoTest` method in `TaskListTest`, that is missing a white space.

Other than that, everything else seems great 👍

javierng2knus reviewed (commented):
LGTM. 👍
Just some very minor coding standard nits to address. 

Edit: 
I realized that we allocated 2 specific people to review instead of just random choosing people...🤦 
Sincerest apologies for any inconvenience caused. 🙇

larainezo committed:
Update JavaFX dependencies in build.gradle

AL-ZT commented on src/main/java/duchess/util/Storage.java:
```diff
@@ -0,0 +1,112 @@
+     * @throws DuchessException if no file found and new file has to be created.
+     */
+    public ArrayList<Task> load() throws DuchessException {
+        ArrayList<Task> list = new ArrayList<>();
```
Perhaps giving an appropriate name to the list of tasks would be better?

AL-ZT commented on src/main/java/duchess/util/Parser.java:
```diff
@@ -0,0 +1,85 @@
+        return details;
+    }
+
+    public String getKeyword(String input) {
```
Maybe the 'getKeyword' method name could be more accurate?

AL-ZT commented on src/main/java/duchess/util/Parser.java:
```diff
@@ -0,0 +1,85 @@
+     * @return int action id.
+     * @throws DuchessException  If no valid action identified.
+     */
+    public int getAction(String input) throws DuchessException {
```
Should this method be named "getActionId"?

AL-ZT commented on src/main/java/duchess/tasks/Task.java:
```diff
@@ -0,0 +1,59 @@
+     *
+     * @return String "X" or " ".
+     */
+    public String getStatusIcon() {
```
I like your method inheritance for task status!

AL-ZT reviewed (commented):
LGTM, just some nits to fix on coding quality.

larainezo committed:
Add JavaFX GUI

larainezo committed:
Add JavaFX GUI

larainezo committed:
Merge pull request #12 from larainezo/Level-10

Level-10

larainezo committed:
Add JavaFX GUI Images

larainezo committed:
Merge pull request #13 from larainezo/A-Varargs

A-Varargs

larainezo committed:
Modify mainClass in build.gradle to create duchess.jar executable

larainezo committed:
getAction() method: check that .txt file is open and loaded

getAction() method does not check whether the "bye" command has been given and the file has already been closed.

Using assert to check if the file has been closed can prevent IOExceptions from happening.

Let's,
* update the getAction() method to use an "assert" statement to check that a file is still open before the user input action is taken

larainezo committed:
Improve Code Quality: Readability and Safe Practices

Code had few empty line breaks, which made readability difficult. Code also had unused import statements.

Using empty line breaks can improve the logical understanding of the code. Removing unused import statements also improves readability and avoids class conflicts.

Let's,
* Add empty line breaks between lines of code with different levels of abstraction
* remove unused import statements

larainezo committed:
A-Streams (Optional) - Not Implemented

larainezo committed:
Merge pull request #14 from larainezo/A-Assertions

A-Assertions

larainezo committed:
Merge branch 'master' into A-CodeQuality

larainezo committed:
Merge pull request #15 from larainezo/A-CodeQuality

A-CodeQuality

larainezo committed:
Merge pull request #16 from larainezo/A-Streams

A-Streams

larainezo committed:
Set up CI using GitHub Actions

larainezo committed:
Merge pull request #17 from larainezo/A-CI

A-CI

larainezo committed:
createTask method: ensure no Task objects with duplicate descriptions are created

New Task objects are always added into taskList regardless of description.

Ensuring that there are no Task objects with duplicate descriptions makes it easier for the user to keep track of upcoming Tasks in the taskList.

Let's,
* Compare the Task description with current descriptions of Task objects in the taskList
* Return an error message when a new Task object is created with a duplicate description

larainezo committed:
Merge pull request #18 from larainezo/C-DetectDuplicates

BCD-Extension

zoebelle-pang commented on src/main/java/duchess/tasks/Event.java:
```diff
@@ -0,0 +1,32 @@
+import java.time.LocalDate;
+import java.time.format.DateTimeFormatter;
+
+public class Event extends Task {
```
Good job on using K&R style brackets throughout your code 👍

zoebelle-pang commented on src/main/java/duchess/util/Parser.java:
```diff
@@ -0,0 +1,206 @@
+     */
+    public Parser() {}
+
+    /**
```
Your use of JavaDoc comments is excellent!

zoebelle-pang commented on src/main/java/duchess/util/TaskList.java:
```diff
@@ -0,0 +1,153 @@
+    public TaskList(ArrayList<Task> taskList) {
+        this.taskList = taskList;
+    }
+
```
Good use of white space throughout your code 😃

zoebelle-pang commented on src/main/java/duchess/util/Parser.java:
```diff
@@ -0,0 +1,206 @@
+            throw new DuchessException(response);
+        }
+
+        return response;
```
Just a tiny suggestion to give the "response" variable a more descriptive name, such as "taskListResponse" or "deleteTaskResponse," to improve readability. Otherwise, your code is well-organized. Good job!

zoebelle-pang reviewed (commented):
Hello! 👋 Great work on your code! It's well-organized and follows Java Coding Standards, with excellent JavaDoc comments. I have only a minor suggestions for some naming improvements. Apart from that, keep up the excellent work! 👍

larainezo committed:
Remove image background

larainezo committed:
Merge pull request #20 from larainezo/A-BetterGui

Remove image background

larainezo committed:
Add title to scene

larainezo committed:
Merge pull request #21 from larainezo/A-Personality

Add title to scene

larainezo committed:
Add pink theme

larainezo committed:
Merge pull request #22 from larainezo/A-Personality

Add pink theme

larainezo committed:
Add README.md

larainezo committed:
Enable assertions in build.gradle

larainezo committed:
Add system exit call
